### ISSUE TITLE
Support distributed reduce on sparse tensors

### ISSUE THREAD

**Issue by mrshenli** [pytorch/pytorch#17369]
## Motivation

This feature request is motivated by #17356 to support sparse tensors in `DistributedDataParallel`. But it might have a broader impact as reducing sparse tensors is not a rare use case.  

## Pitch

Enhance the existing `reduce` API in both [C++](https://github.com/pytorch/pytorch/blob/master/torch/csrc/cuda/nccl.h#L66-L79) and [Python](https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/init.cpp#L202-L224) to support sparse tensors for `sum` and `product` reduce ops (reduce ops are restricted by [available methods in sparse tensors](https://pytorch.org/docs/stable/sparse.html)).  Other distributed communication collectives will follow, but will have their own tracking issues.

**Comment by pietern** [comment]
For distributed, I think the best place to do this is in `ProcessGroup::reduce` and/or `ProcessGroup::allreduce`. It can detect if all input tensors are sparse, and if so, execute a non-standard algorithm. It will have to use send/recv to do the proper merge/reduce needed for sparse tensor reduction, so it would only work for the Gloo and MPI backends. This seems like a good fit for butterfly mixing, since for sparse tensors we need to communicate the entire input tensor, as well as the entire intermediate results anyway.

**Comment by glample** [comment]
Hi, I had a look at the embedding of sparse embedding layers, and it basically look like this:

```python
m = nn.Embedding(10, 3, sparse=True)
m(torch.LongTensor([7, 1, 3])).sum().backward()
print(m.weight.grad)
```
```
torch.sparse.FloatTensor of size (10,3) with indices:
tensor([[7, 1, 3]])
and values:
tensor([[1., 1., 1.],
        [1., 1., 1.],
        [1., 1., 1.]])
```

Is the idea here to use `ProcessGroup::allreduce` to stack the indices / values coming from each worker, just before densifying them and perform the update on each machine?

**Comment by mrshenli** [comment]
@glample if I understand it correctly, the idea is to create a set of communication collectives for sparse tensors, and automatically call them when necessary. We will try not to densify them as there could be super large and super sparse tensors. Instead, we probably will use butterfly mixing to collect indices and values to every worker, and merge them.

**Comment by pietern** [comment]
Will be addressed by #19146 and #19443
